dlrover/trainer/torch/flash_checkpoint/hf_trainer.py

```diff
@@ -0,0 +1,380 @@
+        )
+        os.makedirs(output_dir, exist_ok=True)
+        logger.info(f"Saving model checkpoint to {output_dir}")
+
+        supported_classes = (
+            (PreTrainedModel,)
+            if not is_peft_available()
+            else (PreTrainedModel, PeftModel)
+        )
+        # Save a trained model and configuration using `save_pretrained()`.
+        # They can then be reloaded using `from_pretrained()`
+        if self.args.save_safetensors:
+            logger.warn(
+                "Flash checkpoint does not support safatensors "
+                "and torch.save is used."
+            )
+        if not isinstance(self.model, supported_classes):
+            if state_dict is None:
+                state_dict = self.model.state_dict()
+
+            if isinstance(unwrap_model(self.model), supported_classes):
+                unwrap_model(self.model).save_pretrained(
+                    output_dir,
+                    state_dict=state_dict,
+                    safe_serialization=False,
+                    save_function=self.flash_checkpointer.ckpt_agent.save,
+                )
+            else:
+                logger.info(
+                    "Trainer.model is not a `PreTrainedModel`, "
+                    "only saving its state dict."
+                )
+                torch.save(state_dict, os.path.join(output_dir, WEIGHTS_NAME))
+        else:
+            self.model.save_pretrained(
+                output_dir,
+                state_dict=state_dict,
+                safe_serialization=False,
+                save_function=self.flash_checkpointer.ckpt_agent.save,
+            )
+
+        if self.tokenizer is not None:
+            self.tokenizer.save_pretrained(output_dir)
+
+        args_path = os.path.join(output_dir, TRAINING_ARGS_NAME)
+        self.flash_checkpointer.ckpt_agent.save(self.args, args_path)
+
+    def _rotate_checkpoints(self, use_mtime=False, output_dir=None) -> None:
+        if (
+            self.args.save_total_limit is None
+            or self.args.save_total_limit <= 0
+        ):
+            return
+
+        last_step = self._get_last_checkpoint_step()
+
+        # Check if we should delete older checkpoint(s)
+        checkpoints_sorted = self._sorted_checkpoints(
+            use_mtime=use_mtime, output_dir=output_dir
+        )
+
+        valid_checkpoints = []
+        for path in checkpoints_sorted:
+            regex_match = re.match(f".*{PREFIX_CHECKPOINT_DIR}-([0-9]+)", path)
+            if regex_match is not None and regex_match.groups() is not None:
+                step = int(regex_match.groups()[0])
+                if step <= last_step:
+                    valid_checkpoints.append(path)
+
+        if len(valid_checkpoints) <= self.args.save_total_limit:
+            return
+
+        # If save_total_limit=1 with load_best_model_at_end=True,
+        # we could end up deleting the last checkpoint, which
+        # we don't do to allow resuming.
```

Comment:
`which we don't do  to allow resuming` -> `which should be avoided and allow resuming`